File: solutions/julia/land-grab-in-space/land-grab-in-space.jl
struct Coord
    x::UInt16
    y::UInt16
end

@kwdef struct Plot
    bottom_left::Coord
    top_right::Coord
end

function is_claim_staked(claim::Plot, register::Set{Plot})
    return claim ∈ register
end

function stake_claim!(claim::Plot, register::Set{Plot})
    if is_claim_staked(claim, register)
        return false
    else
        push!(register, claim)
        return true
    end
end

function get_longest_side(claim::Plot)
    return max(claim.top_right.x - claim.bottom_left.x, claim.top_right.y - claim.bottom_left.y)
end

function get_claim_with_longest_side(register::Set{Plot})
    ranking = sort([(c, get_longest_side(c)) for c in register], by = x -> x[2], rev=true)
    return Set(c[1] for c in ranking if c[2] == ranking[1][2])
end

File: solutions/csharp/land-grab-in-space/LandGrabInSpace.cs
public struct Coord
{
    public Coord(ushort x, ushort y)
    {
        X = x;
        Y = y;
    }

    public ushort X { get; }
    public ushort Y { get; }
}

public struct Plot
{
    public Plot(Coord coord0, Coord coord1, Coord coord2, Coord coord3)
    {
        mCoord0 = coord0;
        mCoord1 = coord1;
        mCoord2 = coord2;
        mCoord3 = coord3;
    }
    Coord mCoord0;
    Coord mCoord1;
    Coord mCoord2;
    Coord mCoord3;

    public int Side() => Math.Max(Math.Abs(mCoord0.X - mCoord1.X), Math.Abs(mCoord1.X - mCoord2.Y));
}

public class ClaimsHandler
{
    public void StakeClaim(Plot plot) => claims = [.. claims, plot];
    public bool IsClaimStaked(Plot plot) => claims.Contains(plot);
    public bool IsLastClaim(Plot plot) => plot.Equals(claims[^1]);

    public Plot GetClaimWithLongestSide()
    {
        int longestSideIdx = 0;
        int longestSide = 0;

        for (int i = 0; i < claims.Length; i++)
        {
            if (claims[i].Side() >= longestSide)
            {
                longestSide = claims[i].Side();
                longestSideIdx = i;
            }
        }

        return claims[longestSideIdx];
    }

    Plot[] claims = [];
}
